base_placement_plugin/src/create_marker.cpp

```diff
@@ -0,0 +1,313 @@
+  for (BasePoseJoint::iterator it = baseJoints.begin(); it !=baseJoints.end(); ++it)//Creating each interactive marker
+  {
+    int i=std::distance(baseJoints.begin(), it);
+    bool no_joint_solution = false;
+    geometry_msgs::Pose base_pose = it->second;
+    std::vector<double> joint_soln = it->first;
+    if(std::equal(joint_soln.begin()+1, joint_soln.end(), joint_soln.begin()))
+    {
+          no_joint_solution = true;
+    }
+    moveit::core::RobotStatePtr robot_state_(new moveit::core::RobotState(robot_model_));//Calling robot_state in each iteration
+    for(int k=0;k<joint_soln.size();k++)
+    {
+      robot_state_->setJointPositions(joint_names[k], &(joint_soln[k])); //Setting joint values to arm joints
+    }
+    robot_state_->update(); //Updating robot state
+    const std::vector<std::string>& arm_link_names = robot_jmp->getLinkModelNames();//Obtaining all the arm link names
+    std::vector<std::string> full_link_names = robot_model_->getLinkModelNames(); //Obtaing all the robot model names
+    int position = std::find(full_link_names.begin(), full_link_names.end(), arm_link_names[0]) -full_link_names.begin() ;
+    std::string arm_parent_link = full_link_names[position-1];
+    std::vector<std::string> full_links;
+    for(int k=0;k<arm_link_names.size();k++)
+    {
+      full_links.push_back(arm_link_names[k]);
+    }
+
+    if(has_joint_model_ee)
+    {
+      for(int k=0;k<ee_links.size();k++)
+      {
+        full_links.push_back(ee_links[k]);
+      }
+    }
+
+    visualization_msgs::MarkerArray full_link_markers;
+    robot_state_->getRobotMarkers( full_link_markers, full_links);//Obtain the current markers
+
+    Eigen::Affine3d tf_root_to_first_link = robot_state_->getGlobalLinkTransform(arm_parent_link);//transform of the parent link
+    Eigen::Affine3d tf_first_link_to_root = tf_root_to_first_link.inverse();//inverse of the transform
+
+    Eigen::Affine3d base_tf;
+    tf::poseMsgToEigen(base_pose, base_tf);//base pose in tf
+    visualization_msgs::InteractiveMarker iMarker; //new interactive marker
+    iMarker.header.frame_id = "base_link";
+    iMarker.pose = base_pose;//the pose of the imarker is the pose of the base
+    iMarker.scale = 0.2;
+    std::string name = "robot_model";
+    std::string description = "robot model";
+    iMarker.name = name+boost::lexical_cast<std::string>(i);
+    iMarker.description = description+boost::lexical_cast<std::string>(i);//each imarker has different name and description
+    visualization_msgs::InteractiveMarkerControl robotModelControl;
+    for(std::size_t j=0; j<full_link_markers.markers.size();j++)
+    {
+      full_link_markers.markers[j].header.frame_id = "base_link";
+      full_link_markers.markers[j].type = visualization_msgs::Marker::MESH_RESOURCE;
+      full_link_markers.markers[j].mesh_use_embedded_materials = true;
+      full_link_markers.markers[j].id = i*(j+5);
```

Comment:
What is the +5 for in this line?  Will this be valid for all robot configurations?